File: problem2/problem2.cpp
class FooBar {
private:
    int n;
    mutex m;
    int state = 0;
    condition_variable cv;

    void execute(function<void()> print, int whichTurn) {
        unique_lock<mutex> lck(m);
        for (int i = 0; i < n; i++) {
            cv.wait(lck, [this, whichTurn] { return state == whichTurn; });

            // Run the passed function (printFoo or printBar)
            print();

            // Switch state between foo and bar
            state = 1 - state;
            cv.notify_all();
        }
    }

public:
    FooBar(int n) {
        this->n = n;
    }

    void foo(function<void()> printFoo) {
        execute(printFoo, 0);
    }

    void bar(function<void()> printBar) {
        execute(printBar, 1);
    }
};

File: problem2/problem2.java
class FooBar {
    private int n;
    private int state = 0;

    public FooBar(int n) {
        this.n = n;
    }

    private void execute(Runnable print, int whichTurn) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            synchronized(this) {
                while (state != whichTurn) {
                    wait();
                }
                // Run the passed function (printFoo or printBar)
                print.run();
                // Switch state between foo and bar
                state = 1 - state;
                notifyAll();
            }
        }
    }

    public void foo(Runnable printFoo) throws InterruptedException {
        execute(printFoo, 0);
    }

    public void bar(Runnable printBar) throws InterruptedException {
        execute(printBar, 1);
    }
}
